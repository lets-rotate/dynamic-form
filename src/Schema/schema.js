/* eslint-disable no-prototype-builtins */
import { DynamicFormType } from './type'
import util from '../util'

const schema = (sch) => {
  const expandedSch = {}
  for (const [key, schemaForKey] of Object.entries(sch)) {
    expandedSch[key] = getExpandedSchemaFromKeyAndSchema(key, schemaForKey)
  }

  return expandedSch
}

const getExpandedSchemaFromKeyAndSchema = (key, schemaForKey) => {
  return {
    type: getTypeOfSchemaFromSchema(schemaForKey),
    label: getLabelFromKeyAndSchema(key, schemaForKey),
    placeholder: schemaForKey.placeholder,
    defaultValue: getDefaultValueFromKeyAndSchema(schemaForKey),
    schema: getSubSchemaFromSchema(schemaForKey)
  }
}

const getLabelFromKeyAndSchema = (key, schemaForKey) => {
  if (schemaForKey.label) return schemaForKey.label
  return key
}

const getDefaultValueFromKeyAndSchema = (sch) => {
  const type = getTypeOfSchemaFromSchema(sch)
  if (sch.hasOwnProperty('defaultValue')) {
    const defaultValue = sch.defaultValue
    if (type === DynamicFormType.number) {
      return util.parseInteger(defaultValue)
    }

    return defaultValue
  }

  if (type === DynamicFormType.text && typeof sch === 'string') return sch
  if (type === DynamicFormType.number && typeof sch === 'number') return sch
}

const getTypeOfSchemaFromSchema = (sch) => {
  if (sch.hasOwnProperty('type')) return sch.type

  if (typeof sch === 'string' || sch === String) return DynamicFormType.text
  if (typeof sch === 'number' || sch === Number) return DynamicFormType.number
  if (Array.isArray(sch)) return DynamicFormType.repeatable
  if (typeof sch === 'object') return DynamicFormType.nested
}

const getSubSchemaFromSchema = (sch) => {
  if (sch.hasOwnProperty('schema')) return schema(sch.schema)

  const type = getTypeOfSchemaFromSchema(sch)
  if (type === DynamicFormType.nested) return schema(sch)
  if (type === DynamicFormType.repeatable) return schema(sch[0])
}

export { schema }
