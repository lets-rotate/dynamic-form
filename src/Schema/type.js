export const DynamicFormType = {
  text: 'DynamicFormType.Text',
  number: 'DynamicFormType.Number',
  nested: 'DynamicFormType.Nested',
  repeatable: 'DynamicFormType.Repeatable'
}
