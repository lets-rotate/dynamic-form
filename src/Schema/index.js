export { schema } from 'Schema/schema'
export { DynamicFormSize } from 'Schema/size'
export { DynamicFormType } from 'Schema/type'
