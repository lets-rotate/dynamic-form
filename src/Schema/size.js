export const DynamicFormSize = {
  large: 'DynamicFormSize.Large',
  medium: 'DynamicFormSize.Medium',
  small: 'DynamicFormSize.Small',
  unknown: 'DynamicFormSize.Unknown'
}
