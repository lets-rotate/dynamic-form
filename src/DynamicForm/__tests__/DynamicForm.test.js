/* eslint-disable no-undef */
import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { DynamicForm } from '../index'
import { DynamicFormType } from '../../Schema'

const ROLE_INPUT_STRING = 'role-input-string'
const ROLE_INPUT_DYNAMIC_SUBMIT = 'input-dynamic-submit'

test('DynamicForm with correct props matches snapshot', () => {
  const component = render(
    <DynamicForm
      schema={{
        stringForm: {
          label: 'String Form',
          type: DynamicFormType.text
        }
      }}
      onChange={() => {}}
      onSubmit={() => {}}
    />
  )

  expect(component).toMatchSnapshot()
})

test('DynamicForm will call onChange correctly if form values are changed', (done) => {
  const component = render(
    <DynamicForm
      schema={{
        stringForm: {
          label: 'String Form',
          type: DynamicFormType.text
        }
      }}
      onChange={(v) => {
        expect(v).toMatchObject({
          stringForm: 'New Value'
        })
        done()
      }}
    />
  )

  const stringInput = component.getByRole(ROLE_INPUT_STRING)

  fireEvent.change(stringInput, {
    target: { value: 'New Value' }
  })
})

test('DynamicForm will call onSubmit correctly if submit is tapped', (done) => {
  const component = render(
    <DynamicForm
      schema={{
        stringForm: {
          label: 'String Form',
          type: DynamicFormType.text
        }
      }}
      onSubmit={(v) => {
        expect(v).toMatchObject({
          stringForm: 'New Value'
        })
        done()
      }}
    />
  )

  const stringInput = component.getByRole(ROLE_INPUT_STRING)
  fireEvent.change(stringInput, {
    target: { value: 'New Value' }
  })

  const submitInput = component.getByRole(ROLE_INPUT_DYNAMIC_SUBMIT)
  fireEvent.click(submitInput)
})
