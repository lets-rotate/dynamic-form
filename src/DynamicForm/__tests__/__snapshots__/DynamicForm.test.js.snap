// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicForm with correct props matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="d-grid gap-1"
      >
        <div
          class="card border-light"
          role="component-node"
        >
          <div
            class="card-body p-1 d-grid gap-1"
          >
            <div
              class="input-group flex-nowrap input-group-md"
            >
              <span
                class="input-group-text"
                role="role-label-texty"
              >
                String Form
              </span>
              <input
                class="form-control form-control-md"
                role="role-input-string"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <button
          class="btn btn-success mx-1 w-auto "
          role="input-dynamic-submit"
        >
          Submit
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="d-grid gap-1"
    >
      <div
        class="card border-light"
        role="component-node"
      >
        <div
          class="card-body p-1 d-grid gap-1"
        >
          <div
            class="input-group flex-nowrap input-group-md"
          >
            <span
              class="input-group-text"
              role="role-label-texty"
            >
              String Form
            </span>
            <input
              class="form-control form-control-md"
              role="role-input-string"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <button
        class="btn btn-success mx-1 w-auto "
        role="input-dynamic-submit"
      >
        Submit
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
